### ISSUE TITLE
Pre-commit should install correct python version

### ISSUE THREAD

**Issue by davidhyman** [pre-commit/pre-commit#875]
What?
Pre-commit should install the correct version(s) of python required for running hooks.

Why?
#869 shows `language_version` as the correct way to specify a dependency for a hook. However this isn't acted on by pre-commit at the install stage, so at run time the hook doesn't have that python version available. This violates the purpose of the library:
>Some of the best linters are written in languages that you do not use in your project or have installed on your machine. For example scss-lint is a linter for SCSS written in Ruby. If you’re writing a project in node you should be able to use scss-lint as a pre-commit hook without adding a Gemfile to your project or understanding how to get scss-lint installed.

Notes:
- `ruby.py` contains an `_install_ruby` function, but the equivalent appears missing in `python.py`
- could potentially use [`pyenv`](https://github.com/pyenv/pyenv) to handle the installs

**Comment by asottile** [comment]
Can you show some output of `language_version` not working?  It does as far as I understand:

```yaml
-   repo: local
    hooks:
    -   id: py3
        name: py3
        language: python
        language_version: python3
        entry: python --version
        verbose: true
        pass_filenames: false
        always_run: true
    -   id: py2
        name: py2
        language: python
        language_version: python2
        entry: python --version
        verbose: true
        pass_filenames: false
        always_run: true
    -   id: py35
        name: py35
        language: python
        language_version: python3.5
        entry: python --version
        verbose: true
        pass_filenames: false
        always_run: true
```

```
$ pre-commit  run
py3......................................................................Passed
hookid: py3

Python 3.6.7

py2......................................................................Passed
hookid: py2

Python 2.7.15rc1

py35.....................................................................Passed
hookid: py35

Python 3.5.6

```

**Comment by chriskuehl** [comment]
I think the request is for pre-commit to automatically install versions of Python that you don't have locally on your system, similar to what it does for Node, Ruby, etc.

**Comment by asottile** [comment]
ah I see, well `pyenv` is neither portable nor fast so it's probably not going to happen

For now the [documented requirement](https://pre-commit.com/#overriding-language-version) to use `language_version` for python is to have those available.

**Comment by davidhyman** [comment]
Is speed an important concern? My understanding of `pyenv` is that once it's installed a python, it's ready to be reused if needed. So yes the first-ever run (or `--install-hooks`) is going to take a while, but surely it's better to have the tool handle this for the end user, rather than expect them to go around installing pythons (a notorious minefield).

I have no particular attachment to `pyenv` though, for what it's worth, just the idea that the tool should handle its dependencies as per the stated purpose.

What is the rationale behind supporting the installation of ruby, but not the installation of python?

**Comment by asottile** [comment]
the tool is written in python so python is already available :man_shrugging: -- `docker`, `go`, `rust`, `swift` (and deprecated `pcre` to some extent) currently have the same limitation so it's not just limited to python here.

**Comment by asottile** [comment]
also `pyenv` doesn't even work without also installing another toolchain or two.  For example [it requires an http downloader globally installed](https://github.com/pyenv/pyenv/issues/1009), and requires `libssl-dev` / `gcc` / etc. to be system installed on linuxes.

There aren't great binary distribution methods for python and building from source is a real struggle.  Distributions themselves are much better at providing a working binary

**Comment by davidhyman** [comment]
Agreed, it's not easy :)
Perhaps the versions could be checked at install time and give an explicit error rather than wait until the dependency itself fails.  That would clarify to the user that they are responsible for obtaining the version the hook requires.
It doesn't solve anything much in the grand scheme of things though. Feel free to close this one then if it's out of scope. Thanks for your responses!

**Comment by asottile** [comment]
Do you have an example error message?  The current messaging that I get:

```yaml
-   repo: local
    hooks:
    -   id: py38
        name: py38
        language: python
        language_version: python3.8
        entry: python --version
        verbose: true
        pass_filenames: false
        always_run: true
```

**Comment by ned2** [comment]
Chiming in a bit late to the party as I just ran into this issue with pyenv. 

The error message I got after using `python3.7` as the value for `language_version` was this:

```
pyenv: python3.7: command not found

The `python3.7' command exists in these Python versions:
  3.7.0
  3.7.2
```

It seems as though pyenv has not not registered `python3.7` as a command for my virtualenv (which is based on 3.7.2). Strangely, when I activate the plain `3.7.2` installation, it is registered. 

At any rate, `python3` is a registered command in my virtualenv, so it looks like the value of `python3` `language_version` makes everyone happy.
